File: backend/app/api/ontology.py
"""
List/get/create TargetTypeTemplates - the Target-Type Conceptual
Framework layer described in docs/DTID_ARCHITECTURE.md. Built-ins come
from app.ontology.templates.registry; created templates are held
in-memory only (app.ontology.templates.store) - see that module's
docstring for why.
"""
from fastapi import APIRouter, HTTPException

from app.ontology.schemas import TargetTypeTemplate
from app.ontology.templates.registry import BUILTIN_TEMPLATES, get_builtin_template
from app.ontology.templates.store import (
    CREATED_TEMPLATES,
    add_created_template,
    get_created_template,
    replace_created_template,
)
from app.schemas.ontology import TemplateListResponse, TemplateSummary

router = APIRouter(prefix="/api/v1/ontology", tags=["ontology"])


@router.get("/templates", response_model=TemplateListResponse)
async def list_templates() -> TemplateListResponse:
    all_templates = BUILTIN_TEMPLATES + CREATED_TEMPLATES
    return TemplateListResponse(
        items=[
            TemplateSummary(
                name=t.name,
                description=t.description,
                source=t.source,
                concept_count=len(t.concepts),
            )
            for t in all_templates
        ]
    )


@router.get("/templates/{name}", response_model=TargetTypeTemplate)
async def get_template(name: str) -> TargetTypeTemplate:
    template = get_builtin_template(name) or get_created_template(name)
    if template is None:
        raise HTTPException(status_code=404, detail=f"Template {name!r} not found")
    return template


@router.post("/templates", response_model=TargetTypeTemplate, status_code=201)
async def create_template(template: TargetTypeTemplate) -> TargetTypeTemplate:
    if get_builtin_template(template.name) or get_created_template(template.name):
        raise HTTPException(
            status_code=409, detail=f"A template named {template.name!r} already exists"
        )

    # source is always "user-defined" for anything created through this
    # endpoint, regardless of what the client sent - "builtin" is reserved
    # for the registry in app.ontology.templates.registry.
    created = template.model_copy(update={"source": "user-defined"})
    add_created_template(created)
    return created


@router.put("/templates/{name}", response_model=TargetTypeTemplate)
async def update_template(name: str, template: TargetTypeTemplate) -> TargetTypeTemplate:
    if get_builtin_template(name) is not None:
        raise HTTPException(
            status_code=403, detail=f"Built-in template {name!r} cannot be modified"
        )
    if get_created_template(name) is None:
        raise HTTPException(status_code=404, detail=f"Template {name!r} not found")

    # Name is the identifier and stays fixed for this call - renaming
    # would need a dedicated rename, not a bundled side effect of an edit.
    # source stays "user-defined" regardless of what the client sent, same
    # rule as create_template.
    updated = template.model_copy(update={"name": name, "source": "user-defined"})
    replace_created_template(name, updated)
    return updated

File: frontend/src/types/ontology.ts
export interface ConceptDefinition {
  name: string;
  description: string;
}

export interface ConceptRelation {
  source: string;
  target: string;
}

export interface TargetTypeTemplate {
  name: string;
  description: string;
  source: "builtin" | "user-defined";
  created_by: string | null;
  concepts: ConceptDefinition[];
  relations: ConceptRelation[];
}

export interface TemplateSummary {
  name: string;
  description: string;
  source: "builtin" | "user-defined";
  concept_count: number;
}

export interface TemplateListResponse {
  items: TemplateSummary[];
}

async function parseErrorDetail(response: Response): Promise<string> {
  const detail = await response.json().catch(() => null);
  if (Array.isArray(detail?.detail)) {
    return detail.detail.map((d: { msg?: string }) => d.msg).join("; ");
  }
  return detail?.detail || `Request failed with status ${response.status}`;
}

export async function fetchTemplateList(): Promise<TemplateListResponse> {
  const response = await fetch("/api/v1/ontology/templates");
  if (!response.ok) {
    throw new Error(await parseErrorDetail(response));
  }
  return response.json();
}

export async function fetchTemplate(name: string): Promise<TargetTypeTemplate> {
  const response = await fetch(`/api/v1/ontology/templates/${encodeURIComponent(name)}`);
  if (!response.ok) {
    throw new Error(await parseErrorDetail(response));
  }
  return response.json();
}

export interface CreateTemplatePayload {
  name: string;
  description: string;
  concepts: ConceptDefinition[];
  relations: ConceptRelation[];
}

export async function createTemplate(
  payload: CreateTemplatePayload,
): Promise<TargetTypeTemplate> {
  const response = await fetch("/api/v1/ontology/templates", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!response.ok) {
    throw new Error(await parseErrorDetail(response));
  }
  return response.json();
}

export async function updateTemplate(
  name: string,
  payload: CreateTemplatePayload,
): Promise<TargetTypeTemplate> {
  const response = await fetch(`/api/v1/ontology/templates/${encodeURIComponent(name)}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!response.ok) {
    throw new Error(await parseErrorDetail(response));
  }
  return response.json();
}
